### PR TITLE
[ConstraintSystem] remove unused computation in claimNextNamed

### PR DESCRIPTION
**Pull request by omochi**
This patch removes unused computation in `claimNextNamed`.

See this line in original code.

```c++
        if (argLabel.empty() &&
            (paramInfo.hasDefaultArgument(i) || !forVariadic))
```

I found two issue here.

- A variable `i` passing to `paramInfo` is index not for **parameters** but for **arguments**. So `paramInfo.hasDefaultArgument(i)` may be bug.
- A variable `forVariadic` is always `false` here. So expression `!forVariadic` is always `true` here and previous expression about `paramInfo` is meaningless.

## The reason why `forVariadic` is always `false`

In this `for` loop with `i`, `return` is at bottom.
So to increment `i`, one of three `continue` statement is reached but all are never happens.

About first `continue`, there is `if (forVariadic) return` before `continue`.

About second `continue`, it is in `if (claimedArgs[i])` scope.
It is not entered in first time of loop.
Because `skipClaimedArgs()` before `for` loop seek `nextArgIdx` to index of first unclaimed argument.

About third `continue`, it is in `if (i != nextArgIdx)`.
It is not entered in first time of loop.

In conclusion, when `forVariadic` is true,
this `for` loop is not repeated more than once and always reaches one of two `return` in first time of loop.

## Patch content

With above conclusion, it turns out that `forVariadic` is always `false` also in `if (claimedArgs[i])` scope.

So I write this patch as:

- add assertion about `forVariadic` in `if (claimedArgs[i])` as hint for code reader in future
- add assertion about `forVariadic` in `if (i != nextArgIdx)` as hint for code reader in future
- remove unused expression (`(paramInfo.hasDefaultArgument(i) || !forVariadic)`)

### CODE REVIEW

**Comment by omochi** [comment]
@xedin Please review this PR 🙂

**Comment by xedin** [comment]
@omochi Thank you! I'll try to get to this today.

File: lib/Sema/CSSimplify.cpp [hunk 1]
```diff
@@ -313,6 +313,7 @@ matchCallArguments(SmallVectorImpl<AnyFunctionType::Param> &args,
 
       // Skip claimed arguments.
       if (claimedArgs[i]) {
+        assert(!forVariadic && "Never here when for variadic");
```
[the remark this answers is not in the document]
Comment (omochi) [reply lib/Sema/CSSimplify.cpp]: OK. I will change it.

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -330,8 +332,7 @@ matchCallArguments(SmallVectorImpl<AnyFunctionType::Param> &args,
         // func foo(_ a: Int, _ b: Int = 0, c: Int = 0, _ d: Int) {}
         // foo(1, c: 2, 3) // -> `3` will be claimed as '_ b:'.
         // ```
-        if (argLabel.empty() &&
-            (paramInfo.hasDefaultArgument(i) || !forVariadic))
+        if (argLabel.empty())
```
[the remark this answers is not in the document]
Comment (omochi) [reply lib/Sema/CSSimplify.cpp]: I will add tests.

**Comment by omochi** [comment]
I added test cases about Swift code referred from comment between `if (i != nextArgIdx) {` and `if (argLabel.empty())`.
I wrote 5 functions.
One is about unlabeled parameters.
Others are about variadics parameters.

As whether label of variadics exists or not and whether position of variadics is before or after labeled parameter, I introduced 4 variations.

Each 5 functions can be considered many incorrect and correct invocations.
So test cases became a too lot.

I am wondering if I am on right direction 🤔

**Review by xedin — APPROVED** [review]
Looks great! Thank you for adding so many tests!

**Comment by xedin** [comment]
The more tests we have the better, especially so for `matchCallArguments`! :)

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Looks like some of the new tests failed on Linux...

**Comment by omochi** [comment]
I'll look into test result on Linux

**Comment by omochi** [comment]
I found out that test failure comes from change of diagnostics when I rebase patch onto master.
I will fix it.

**Comment by omochi** [comment]
@xedin I fixed it. Please run test bot.
If it's all ok, I will squash commits into one.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@omochi Seems like this type everything is green.

**Comment by omochi** [comment]
I squashed them into one.

**Comment by xedin** [comment]
Ok, thank you let's get this merged!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Thank you again!
